### DISCUSSION TITLE
Preconditioners for FE INS simulations

### DISCUSSION THREAD

**Discussion by Eilloo** [idaholab/moose#23319] [Q&A Modules: Navier-Stokes] [answered]
Hello,

I've been trying to run INS simulations with various open flow type geometries. For instance, flow around a circular cylinder, or flow over a flat plate. 
Using the `lu` preconditioner, I am able to run simulations even at high Reynolds numbers (I'm wanting to experiment up to around Re~1e6). 
However, even at much lower Reynolds numbers, I have been unable to achieve convergence of the linear solves using any preconditioner besides `lu`. 

I am hoping to simulate some large meshes in future and I gather that whilst robust, lu is very memory intensive. Therefore, I would like to have the option to use some other preconditioner better suited to large problems.

Generally speaking, I have seen no difference between steady and transient solves. 
For transient solves, I have also tried ramping the viscosity so that the simulation starts with a low Reynolds number, then gradually increasing it with time. 
Whilst the residual starts off smaller, the convergence appears to be no better even at the first time step.

Worth mentioning among the options I have tried so far, I noticed that adjusting `pc_factor_levels` with `ilu` did not seem to make a difference. 
However, from my understanding, the more 'levels of fill' with ilu, the closer its resemblance to lu? 
I would therefore expect to see some difference, and wonder if I am misunderstanding something about what ilu is doing.

I have also read through #16113 and tried:
```
petsc_options_iname = '-pc_type -pc_asm_overlap -sub_pc_type -sub_pc_factor_levels  -ksp_gmres_restart'
petsc_options_value = 'bjacobi      2               ilu          4       200'
```
and a few variations, but no avail. 

In my related thread #23274, GiudGiud suggests that something involving FSP could be helpful here; however, in #16113, lindsayad mentions that he's found asm and bjacobi to be better for INS problems.

Any advice on where I should look here would be greatly appreciated. Choosing appropriate preconditioners still seems something of a mystery to me at times!

For what it's worth, below is an example input file in case there are other mistakes I'm making!

Thanks!

```
filename = exampleFile
velocity_x = 5 # 5 for Re=5million at x=1

[Mesh]
    [base]
        type = FileMeshGenerator
        file = ./mesh/FlatPlateMesh_2D.msh
    []
[]

[Variables]
    [velocity]
        family = LAGRANGE_VEC
        block = 'fluid'
    []
    [pressure]
        block = 'fluid'
    []
[]

[ICs]
    [velocity]
        type = VectorConstantIC
        x_value = ${velocity_x}
        y_value = 1e-13
        variable = velocity
    []
[]

[Functions]
    # [mu]                         #Use this to try ramping Re in transient solves
    #     type = ParsedFunction
    #     value = 'if (t <= 0.5, 1.789e-5 - ((t-0.1)*(1.789 - 1.789e-5)/0.1), 1.789e-5)'    #Ramps from 1.7e-2 to 1.7e-5
    # []
    [mu]                           #Fixed mu for SS solves
        type = ParsedFunction
        value = '1.789e-5'
    []
[]

[Materials]
    [fluid] # Air, ISA sea lvl
        type = ADGenericFunctionMaterial
        prop_names = 'rho mu T cp k'
        prop_values = '1.225 mu 288 1005 0.02596'
        block = 'fluid'
    []
    [ins_mat]
        type = INSADStabilized3Eqn
        velocity = velocity
        pressure = pressure
        temperature = '288'
        block = 'fluid'
        alpha = 1.0
    []
[]

[Kernels]
    [mass_conservation]
        type = INSADMass
        variable = pressure
    []
    # [momentum_time]
    #     type = INSADMomentumTimeDerivative
    #     variable = velocity
    # []
    [momentum_advection]
        type = INSADMomentumAdvection
        variable = velocity
    []
    [momentum_pressure]
        type = INSADMomentumPressure
        variable = velocity
        pressure = pressure
    []
    [momentum_viscous]
        type = INSADMomentumViscous
        variable = velocity
    []

    [momentum_supg]
        type = INSADMomentumSUPG
        variable = velocity
        velocity = velocity
    []
    [mass_pspg]
        type = INSADMassPSPG
        variable = pressure
    []
[]

[BCs]
    [inlet]
        type = VectorDirichletBC
        values = '${velocity_x} 0 0'
        variable = velocity
        boundary = 'inlet'
    []
    [outlet]
        type = ADDirichletBC
        value = 0
        variable = pressure
        boundary = 'outlet'
    []
    [free_slip]
        type = ADVectorFunctionDirichletBC
        variable = velocity
        boundary = 'bottom_upstream top'
        function_y = 0
        set_x_comp = False
        set_y_comp = True
        set_z_comp = False
    []
    [no_slip_wall]
        type = VectorDirichletBC
        values = '0 0 0'
        variable = velocity
        boundary = 'plate'
    []
[]

[Preconditioning]
    [precond]
        type = SMP
        full = true
    []
[]

[Executioner]
    type = Steady
    # dt = 0.0025
    # end_time = 1
    solve_type = NEWTON
    nl_rel_tol = 1e-10
    nl_max_its = 35
    l_tol = 1e-10
    l_max_its = 50
    automatic_scaling = true
    line_search = default

    #These petsc options work:
    petsc_options_iname = '-pc_type -pc_factor_shift_type'      
    petsc_options_value = 'lu       NONZERO'


    #Uncomment below for transient solves

    # dtmax = 0.0025
    # [TimeStepper]
    #     type = IterationAdaptiveDT
    #     # dt = 1E-5
    #     dt = 0.0001
    #     growth_factor = 1.5
    #     cutback_factor = 0.8
    # []
[]

[Outputs]
    [OutputFile]
        type = Exodus
        file_base = ${filename}
    []
[]
```

**Comment by GiudGiud** [comment]
Have you tried field splits?

**Reply by Eilloo** [reply]
Not yet, as I believe vector variables are not currently accommodated? However, to use the momentum SUPG kernel, I am required to pass a vector variable for velocity.
I am going to try rewriting the input file with separate variables for u and v, perhaps without SUPG or perhaps using aux variables to convert between vectors and standard variables. 
This should enable me to try field split, and I will report back.

**Reply by GiudGiud** [reply]
We have added more support for field splits with vector variables about two months ago.

AuxVariables for conversions likely wont work. Auxvariables wont keep track of the dependencies of residual terms on variables

What flow regime is this? What s the Reynolds number? You might have to use SUPG

**Reply by Eilloo** [reply]
Reynolds number wise, I am going as high as e6 (I will be experimenting with turbulence options when Re is this high), so I suspect SUPG will be needed?

Ah, good to hear - I'll make sure I'm up-to-date. With the new support, are we able to split a vector variable into its constituent components for the field split? 
Or would we be treating the vector as one variable, and preconditioning this (ie, all 3 components) in one split, and other variables such as pressure in another?

If the case is the former, I have been struggling to find how to - or indeed, if it is possible to - refer to a specific component of a vector variable from the MOOSE input file.

**Reply by GiudGiud** [reply]
You will need SUPG

No you wont be able to split vector variables by component. You can take a look at test/tests/preconditioning/fsp/vector-test.i for an example.

**Reply by Eilloo** [reply]
Thanks for confirming - I'll try splitting my velocity vector and pressure variables and see if things improve

**Reply by GiudGiud** [reply]
You can have a look at this PR on the same topic but just finite volume
https://github.com/idaholab/moose/pull/22405

**Reply by maxnezdyur** [reply]
@Eilloo This might work for you. Let me know if it speeds up the problem!
```
[FSP]
    type = FSP
    topsplit = 'all'
    [all]
      splitting = 'mom mass'
      splitting_type = schur
      schur_type = full
      petsc_options_iname = '-pc_fieldsplit_schur_fact_type -pc_fieldsplit_schur_precondition'
      petsc_options_value = 'full selfp '
    []
    [mom]
      vars = 'velocity'
      petsc_options_iname = '-ksp_type -pc_type -sub_pc_type -sub_ksp_type'
      petsc_options_value = ' preonly  asm      jacobi          preonly'
    []
    [mass]
      vars = 'pressure'
      petsc_options_iname = '-ksp_type -pc_type -pc_hypre_type '
      petsc_options_value = ' preonly   hypre  boomeramg'
    []
  []
```

**Reply by Eilloo** [reply, marked as answer]
An update on this: the good news is that I am getting convergence with some field split options:

What I've tried is far from exhaustive, but so far the following basic options are working okay:
`splitting_type = multiplicative`
and for both pressure, and velocity separately:
`-pc-type = ilu`
It's not the fastest, taking around 80 linear iterations each nl solve, but it is converging without lu which is progress.

Max, I did try the options suggested above, but this diverged at the first iteration for me.

As for the settings which worked, these are fairly arbitrary currently, and I'm trying to get a better idea of what's actually going on.
The splitting type certainly makes a big difference - 'additive', for instance does not work nearly as well. I haven't found much detail in the docs on what the differences are so this has been trial and error so far...

Many of the suggestions, including Max's above, involve Schur's complement. From what I have read in the Petsc docs, it seems that whilst there are built-in methods, preconditioning of the Schur's complement matrix itself is non-trivial and often problem specific? 

Another general observation which I'm not sure I was expecting, is that the convergence is completely different using FSP with, say, both variables preconditioned with ilu, compared to straight ilu with no field split.
Is this because the systems are decoupled in some way? (Perhaps 'decoupled' is the wrong word here - what I'm getting at is that it's hard to picture how two coupled variables are being isolated in separately preconditioned systems)

After looking at the FV bits linked to, I wonder if you can clarify one more thing for me:
What is the SMP preconditioner acting on? The fact we define both the SMP block, then separately some other preconditioners suggestes they operate on different systems, and I'm interested to know roughly what's going on here.

**Reply by GiudGiud** [reply]
This is the documentation on field splits
https://petsc.org/release/docs/manualpages/PC/PCFIELDSPLIT/

You shoud be able to map additive, multiplicative and Schur to which term in the coupling is being considered.
